Source/KSCrash/Source/KSCrash/Recording/BSG_KSMachHeaders.m

```diff
@@ -0,0 +1,190 @@
+            }
+            break;
+        }
+        case LC_UUID: {
+            struct uuid_command *uuidCmd = (struct uuid_command *)cmdPtr;
+            uuid = uuidCmd->uuid;
+            break;
+        }
+        }
+        cmdPtr += loadCmd->cmdsize;
+    }
+    
+    // Save these values
+    info->mh = header;
+    info->cpusubtype = header->cpusubtype;
+    info->cputype = header->cputype;
+    info->imageSize = imageSize;
+    info->imageVmAddr = imageVmAddr;
+    info->uuid = uuid;
+    info->name = image_name;
+    
+    return true;
+}
+
+/**
+ * A callback invoked when dyld loads binary images.  It stores enough relevant info about the
+ * image to populate a crash report later.
+ *
+ * @param header A mach_header structure
+ *
+ * @param slide The VM offset of the binary image
+ */
+void bsg_mach_binary_image_added(const struct mach_header *header, intptr_t slide)
+{
+    BSG_Mach_Binary_Image_Info info = { 0 };
+    if (populate_info(header, &info)) {
+        [bsg_mach_binary_images_info addObject:[NSValue valueWithBytes:&info
+                                                          objCType:@encode(BSG_Mach_Binary_Image_Info)]];
+    }
+}
+
+/**
+ * Compare two mach biunary image structures
+ */
+bool infos_are_equal(BSG_Mach_Binary_Image_Info *info1, BSG_Mach_Binary_Image_Info *info2) {
+    return info1->cpusubtype == info2->cpusubtype &&
+        info1->cputype == info2->cputype &&
+        info1->imageSize == info2->imageSize &&
+        info1->imageVmAddr == info2->imageVmAddr &&
+        info1->name == info2->name &&
+        info1->uuid == info2->uuid &&
+        info1->mh->filetype == info2->mh->filetype &&
+        info1->mh->flags == info2->mh->flags &&
+        info1->mh->magic == info2->mh->magic &&
+        info1->mh->ncmds == info2->mh->ncmds &&
+        info1->mh->sizeofcmds == info2->mh->sizeofcmds;
+}
+
+/**
+ * Called when a binary image is unloaded.
+ */
+void bsg_mach_binary_image_removed(const struct mach_header *header, intptr_t slide)
+{
+    // Convert header and slide into an info struct
+    BSG_Mach_Binary_Image_Info info;
+    if (populate_info(header, &info)) {
+        
+        // We need to search the array manually.  This should be an infrequent operation.
+        for (int i=0;i<[bsg_mach_binary_images_info count];++i) {
+            
+            // Reconstitute the struct from the stored NSValue
+            BSG_Mach_Binary_Image_Info array_info;
+            [[bsg_mach_binary_images_info objectAtIndex:i] getValue:&array_info];
+            
+            // Remove it if it's been found.  Can't have a binary imag eloaded more than once, so safe to break.
+            if (infos_are_equal(&array_info, &info)) {
+                [bsg_mach_binary_images_info removeObjectAtIndex:i];
+                break;
+            }
+        }
+    }
+}
+
+/**
+ * Returns a C array of structs describing the loaded Mach binaries
+ */
+BSG_Mach_Binary_Image_Info* bsg_mach_header_array(size_t *count) {
```

Comment:
My gut feel is that this function could be done away with if the structs are stored as a C array